### DISCUSSION TITLE
graph: Add method to create edges in bulk

### DISCUSSION THREAD

**Discussion by Pzoom522** [networkit/networkit#680] [Q&A] [answered]
In `networkx`, we can create a new graph with **a batch of edges**, e.g., `G.add_edges_from([(1,2),(1,3)])`. However, in `networkit` I only find `addEdge()` which can only add **one edge each time**. For I/O efficiency, is it possible to create a (directed) graph using an edge list with existing APIs? Many Thanks!

**Comment by DeepaMahm** [comment]
@Pzoom522  Does this help 
```
# Networkx graph to NetworKit graph
G_nk = nk.nxadapter.nx2nk(G)
```
?

**Comment by Pzoom522** [comment]
> @Pzoom522 Does this help
> 
> ```
> # Networkx graph to NetworKit graph
> G_nk = nk.nxadapter.nx2nk(G)
> ```
> 
> ?

Thanks a lot for your reply! This API works. However, I just found that the efficiency of `networkx`'s graph construction is quite low. Therefore, It would be better if naive `networkit` APIs could build a graph using a list of tuples...

**Comment by avdgrinten** [comment]
I guess such an API would be reasonable but it doesn't exist yet. Do you need to parse a custom graph format in Python?

**Comment by Pzoom522** [comment]
> I guess such an API would be reasonable but it doesn't exist yet. Do you need to parse a custom graph format in Python?

Exactly! To start with, I think it would be great if we can have a parser which allows custom `HEADER`, `SEPARATOR`, `DELIMITER`, and `FOOTER`. It will be very scalable, e.g., [the space separated edgelist](https://github.com/networkit/networkit/blob/master/input/spaceseparated.edgelist) then boils down to a special case where `HEADER = ""`, `SEPARATOR = " "`, `DELIMITER = "\n"`, and `FOOTER = ""`.

**Comment by Pzoom522** [comment]
In addition, it seems to me the the I/O of `networkit` has not been parallelised, which can be a little bit slow when handling large graphs?

**Comment by avdgrinten** [comment, marked as answer]
It's true that we do not parallelize I/O for most graph formats. In most cases, that would be somewhat hard to do.

One notable exception is the NetworKit binary format. It can be read much faster than other formats since it was designed with parallelism in mind. If you often need to read the same graphs, you could store a copy of them in NetworKit's binary format. (But of course, that does not help if you need to generate the data dynamically...)

**Comment by Pzoom522** [comment]
> It's true that we do not parallelize I/O for most graph formats. In most cases, that would be somewhat hard to do.
> 
> One notable exception is the NetworKit binary format. It can be read much faster than other formats since it was designed with parallelism in mind. If you often need to read the same graphs, you could store a copy of them in NetworKit's binary format. (But of course, that does not help if you need to generate the data dynamically...)

Thanks for you reply! Consider this pipeline: (1) split a large edgelist file into several pieces, (2) load them to init a group of small graphs, (3) use `merge` to merge them into a large graph. If we conduct this in a parallel fashion (e.g., mapreduce), do you think it can substantially shorten loading process?

BTW, are there any APIs which support merging or appending a set of graphs (in batchs / in parallel)?

**Comment by hmeyerhenke** [comment]
Ensuring consistency in the graph data structure in phase (3) would eat away some of the parallelism you gained in phase (2). It's in principle possible, but we recommend the binary format if you want the best performance.

**Comment by Pzoom522** [comment]
> Ensuring consistency in the graph data structure in phase (3) would eat away some of the parallelism you gained in phase (2). It's in principle possible, but we recommend the binary format if you want the best performance.

Thanks a lot. However, the primary issue here is that the large graphs to be proceed are **dynamic** and are given as edge lists, and each is only needed to be load once. Therefore, performing the format conversion seems unhelpful. Regrading (3), do you know are there any APIs which support merging or appending a batch of graphs in a parallel manner?

**Comment by hmeyerhenke** [comment]
Parallel coarsening in PLM (parallel Louvain method, community module) creates thread-local coarse graphs which are then merged. Conceptually, this could be a blueprint for what you have in mind.

**Comment by Pzoom522** [comment]
> Parallel coarsening in PLM (parallel Louvain method, community module) creates thread-local coarse graphs which are then merged. Conceptually, this could be a blueprint for what you have in mind.

Many thanks!
